Panic on parsing malformed dockerfile (2)

Thanks for fixing https://github.com/uber/makisu/issues/266!

**Describe the bug**

I've found a couple of more panics by [go-fuzz](https://github.com/dvyukov/go-fuzz):

**To Reproduce**
1. `FROM 0\nCOPY \xa0`:

```go
// +build ignore

package main

import (
	"github.com/uber/makisu/lib/parser/dockerfile"
)

func main() {
	data := string([]byte("FROM 0\nCOPY \xa0"))

	dockerfile.ParseFile(data, map[string]string{})
}

```

fails with
```
panic: runtime error: index out of range [0] with length 0

goroutine 1 [running]:
github.com/uber/makisu/lib/parser/dockerfile.newCopyDirective(0xc00009a2a0, 0xc00009a1e0, 0xc0000980dc, 0x4, 0xc00009e860, 0xc000098001)
	/Users/bayandin/go/src/github.com/uber/makisu/lib/parser/dockerfile/copy.go:39 +0x6a3
github.com/uber/makisu/lib/parser/dockerfile.newDirective(0xc0000980d4, 0x6, 0xc00009a1e0, 0xff9, 0xc0000980d4, 0x6, 0x0)
	/Users/bayandin/go/src/github.com/uber/makisu/lib/parser/dockerfile/directive.go:59 +0x1be
github.com/uber/makisu/lib/parser/dockerfile.ParseFile(0xc000098002, 0xd, 0xc00009a180, 0xd, 0xc000098002, 0xd, 0xc000016118, 0x0)
	/Users/bayandin/go/src/github.com/uber/makisu/lib/parser/dockerfile/parse_file.go:42 +0x2b7
main.main()
	/Users/bayandin/go/src/github.com/uber/makisu/lib/parser/dockerfile/test.go:12 +0xa2
exit status 2
```

2.  `FROM \x85`

```go
// +build ignore

package main

import (
	"github.com/uber/makisu/lib/parser/dockerfile"
)

func main() {
	data := string([]byte("FROM \x85"))

	dockerfile.ParseFile(data, map[string]string{})
}

```

fails with 

```
panic: runtime error: index out of range [0] with length 0

goroutine 1 [running]:
github.com/uber/makisu/lib/parser/dockerfile.newFromDirective(0xc000076360, 0xc000076330, 0xc00009e088, 0x4, 0xc000085680, 0xc00009e001)
	/Users/bayandin/go/src/github.com/uber/makisu/lib/parser/dockerfile/from.go:49 +0x30f
github.com/uber/makisu/lib/parser/dockerfile.newDirective(0xc00009e080, 0x6, 0xc000076330, 0x1000, 0xc00009e080, 0x6, 0xffffffffffffffff)
	/Users/bayandin/go/src/github.com/uber/makisu/lib/parser/dockerfile/directive.go:59 +0x1be
github.com/uber/makisu/lib/parser/dockerfile.ParseFile(0xc00009e002, 0x6, 0xc0000762d0, 0x6, 0xc00009e002, 0x6, 0x0, 0x0)
	/Users/bayandin/go/src/github.com/uber/makisu/lib/parser/dockerfile/parse_file.go:42 +0x2b7
main.main()
	/Users/bayandin/go/src/github.com/uber/makisu/lib/parser/dockerfile/test.go:12 +0x93
exit status 2
```

**Expected behaviour**
A proper error message that dockerfile is malformed

**Additional context**
makisu version is 27daadc

--- a/lib/parser/dockerfile/add.go
+++ b/lib/parser/dockerfile/add.go
@@ -27,7 +27,12 @@ func newAddDirective(base *baseDirective, state *parsingState) (Directive, error
 	if err := base.replaceVarsCurrStage(state); err != nil {
 		return nil, err
 	}
-	d, err := newAddCopyDirective(base, strings.Fields(base.Args))
+	args := strings.Fields(base.Args)
+	if len(args) == 0 {
+		return nil, base.err(errMissingArgs)
+	}
+
+	d, err := newAddCopyDirective(base, args)
 	if err != nil {
 		return nil, err
 	}

--- a/lib/parser/dockerfile/base.go
+++ b/lib/parser/dockerfile/base.go
@@ -15,7 +15,6 @@
 package dockerfile
 
 import (
-	"errors"
 	"fmt"
 	"regexp"
 	"strings"
@@ -70,9 +69,6 @@ func (d *baseDirective) replaceVars(vars map[string]string) error {
 	replaced, err := replaceVariables(d.Args, vars)
 	if err != nil {
 		return d.err(fmt.Errorf("Failed to replace variables in input: %s", err))
-	}
-	if len(replaced) == 0 {
-		return d.err(errors.New("Empty args after replacing variables"))
 	}
 	d.Args = replaced
 	return nil

--- a/lib/parser/dockerfile/copy.go
+++ b/lib/parser/dockerfile/copy.go
@@ -34,6 +34,9 @@ func newCopyDirective(base *baseDirective, state *parsingState) (Directive, erro
 		return nil, err
 	}
 	args := strings.Fields(base.Args)
+	if len(args) == 0 {
+		return nil, base.err(errMissingArgs)
+	}
 
 	var fromStage string
 	if val, ok, err := parseFlag(args[0], "from"); err != nil {

--- a/lib/parser/dockerfile/expose.go
+++ b/lib/parser/dockerfile/expose.go
@@ -32,7 +32,12 @@ func newExposeDirective(base *baseDirective, state *parsingState) (Directive, er
 	if err := base.replaceVarsCurrStage(state); err != nil {
 		return nil, err
 	}
-	return &ExposeDirective{base, strings.Fields(base.Args)}, nil
+	args := strings.Fields(base.Args)
+	if len(args) == 0 {
+		return nil, base.err(errMissingArgs)
+	}
+
+	return &ExposeDirective{base, args}, nil
 }
 
 // Add this command to the build stage.

--- a/lib/parser/dockerfile/from.go
+++ b/lib/parser/dockerfile/from.go
@@ -37,6 +37,9 @@ func newFromDirective(base *baseDirective, state *parsingState) (Directive, erro
 		return nil, err
 	}
 	args := strings.Fields(base.Args)
+	if len(args) == 0 {
+		return nil, base.err(errMissingArgs)
+	}
 
 	var alias string
 	if len(args) > 1 {

--- a/lib/parser/dockerfile/from_test.go
+++ b/lib/parser/dockerfile/from_test.go
@@ -31,6 +31,7 @@ func TestNewFromDirective(t *testing.T) {
 		image   string
 		alias   string
 	}{
+		{"missing args", false, "from ", "", ""},
 		{"no tag, no alias", true, "from test_image", "test_image", ""},
 		{"tag, no alias", true, "from test_image:trusty", "test_image:trusty", ""},
 		{"no tag, alias", true, "from test_image as test_alias", "test_image", "test_alias"},

--- a/lib/parser/dockerfile/volume.go
+++ b/lib/parser/dockerfile/volume.go
@@ -36,8 +36,12 @@ func newVolumeDirective(base *baseDirective, state *parsingState) (Directive, er
 	if volumes, ok := parseJSONArray(base.Args); ok {
 		return &VolumeDirective{base, volumes}, nil
 	}
+	args := strings.Fields(base.Args)
+	if len(args) == 0 {
+		return nil, base.err(errMissingArgs)
+	}
 
-	return &VolumeDirective{base, strings.Fields(base.Args)}, nil
+	return &VolumeDirective{base, args}, nil
 }
 
 // Add this command to the build stage.

